proposals/intraprocess_communication.md

```diff
@@ -0,0 +1,580 @@
+   receives one if set to transient local).
+   The `rcl_guard_condition_t` member of `SubscriptionIntraProcessWaitable` of each `Subscription` is triggered (this wakes up
+   `rclcpp::spin`).
+
+The way in which the `std::unique_ptr<MessageT>` message is "added" to a buffer, depends on the type of the buffer.
+
+ - `BufferT = unique_ptr<MessageT>`
+   The buffer receives a copy of `MessageT` and has ownership on it. For the last buffer, a copy is not
+   necessary as ownership can be transferred.
+ - `BufferT = shared_ptr<const MessageT>`
+   Every buffer receives a shared pointer of the same `MessageT`. No copies are required.
+ - `BufferT = MessageT`
+   A copy of the message is added to every buffer.
+
+
+#### Publishing other message types
+
+The `Publisher::publish(...)` method is overloaded to support different message types.
+
+ - `unique_ptr<MessageT>`
+ - `MessageT &`
+ - `MessageT*`
+ - `const shared_ptr<const MessageT>`
+
+The last two of them are actually deprecated since ROS2 Dashing.
+All these methods are unchanged with respect to the current implementation: they end up creating a `unique_ptr`
+and calling the `Publisher::publish(std::unique_ptr<MessageT> msg)` described above.
+
+### Receiving intra-process messages
+
+As previously described, whenever messages are added to the ring buffer of a `Subscription`, a condition
+variable specific to the `Subscription` is triggered. This condition variable has been added to the `Node` waitset
+so it is being monitored by the `rclcpp::spin`.
+
+Remember that the `SubscriptionIntraProcessWaitable` object has access to the ring buffer and to the callback function pointer of
+its related `Subscription`.
+
+1. The guard condition linked with the `SubscriptionIntraProcessWaitable` object awakes `rclcpp::spin`.
+2. The `SubscriptionIntraProcessWaitable::is_ready()` condition is checked. This has to ensure that the ring buffer is not empty.
+3. The `SubscriptionIntraProcessWaitable::execute()` function is triggered. Here the first message is extracted from the buffer and
+   then the `SubscriptionIntraProcessWaitable` calls the `AnySubscriptionCallback::dispatch_intra_process(...)` method.
+   There are different implementations for this method, depending on the data-type stored in the buffer.
+4. The `AnySubscriptionCallback::dispatch_intra_process(...)` method triggers the associated callback.
+   Note that in this step, if the type of the buffer is a smart pointer one, no message copies occurr, as
+   ownership has been already taken into account when pushing a message into the queue.
+
+
+### Handling intra and inter-process communications
+
+### Setting up the middleware
+
+In order to avoid the issue previously described, i.e., that a `Subscription` would receive both inter and
+intra-process messages, it is necessary to inform the underlying middleware that some "connections" will be
+handled at the `rclcpp` level, enabling the RMW to neglect them.
```

Comment:
This functionality could be implemented independently of the separate subscriber-specific buffers to address the problem described in scenario 1. I think the proposal should separate this into two independent changes.